File: lib/java/src/main/java/uapi/internal/SerializeInternal.java
package uapi.internal;

import java.util.List;
import java.util.Objects;

import uapi.Message;
import uapi.SerializationError;
import uapi.Serialization;
import uapi.internal.binary.BinaryEncoder;
import uapi.internal.binary.BinaryEncoderUnavailableError;

public class SerializeInternal {
    public static byte[] serializeInternal(Message message, BinaryEncoder binaryEncoder,
            Serialization serializer) {
        final var headers = message.header;

        final boolean serializeAsBinary;
        if (headers.containsKey("_binary")) {
            serializeAsBinary = Objects.equals(true, headers.remove("_binary"));
        } else {
            serializeAsBinary = false;
        }

        final List<Object> messageAsPseudoJson = List.of(message.header, message.body);

        try {
            if (serializeAsBinary) {
                try {
                    final var encodedMessage = binaryEncoder.encode(messageAsPseudoJson);
                    return serializer.toMsgPack(encodedMessage);
                } catch (BinaryEncoderUnavailableError e) {
                    // We can still submit as json
                    return serializer.toJson(messageAsPseudoJson);
                }
            } else {
                return serializer.toJson(messageAsPseudoJson);
            }
        } catch (Throwable e) {
            throw new SerializationError(e);
        }
    }
}

File: lib/ts/uapi/internal/SerializeInternal.ts
import { SerializationError } from 'uapi/SerializationError';
import { Serialization } from 'uapi/Serialization';
import { Message } from 'uapi/Message';
import { BinaryEncoder } from 'uapi/internal/binary/BinaryEncoder';

export function serializeInternal(
    message: Message,
    binaryEncoder: BinaryEncoder,
    serializer: Serialization,
): Uint8Array {
    const headers: Record<string, any> = message.header;

    let serializeAsBinary: boolean;
    if ('_binary' in headers) {
        serializeAsBinary = headers['_binary'] === true;
        delete headers['_binary'];
    } else {
        serializeAsBinary = false;
    }

    const messageAsPseudoJson: any[] = [message.header, message.body];

    try {
        if (serializeAsBinary) {
            try {
                const encodedMessage = binaryEncoder.encode(messageAsPseudoJson);
                return serializer.toMsgpack(encodedMessage);
            } catch (error) {
                // We can still submit as JSON
                return serializer.toJson(messageAsPseudoJson);
            }
        } else {
            return serializer.toJson(messageAsPseudoJson);
        }
    } catch (error) {
        throw new SerializationError(error as Error);
    }
}
